vault/logical_system_test.go

```diff
@@ -2058,6 +2058,96 @@ func TestSystemBackend_tuneAuth(t *testing.T) {
+	req.Data["user_lockout_config"] = map[string]interface{}{
+		"lockout_threshold": "3",
+	}
+	req.ClientToken = rootToken
+	resp, err := b.HandleRequest(namespace.RootContext(nil), req)
+	if err == nil || resp == nil || !resp.IsError() || !strings.Contains(resp.Error().Error(), "tuning of user lockout configuration for auth type \"token\" not allowed") {
+		t.Fatalf("expected tune request to fail, but got resp: %#v, err: %s", resp, err)
+	}
+
+	// tune user_lockout_config for userpass auth type
+	userpassMe := &MountEntry{
+		Table:       credentialTableType,
+		Path:        "userpass/",
+		Type:        "userpass",
+		Description: "userpass",
+	}
+	err = c.enableCredential(namespace.RootContext(nil), userpassMe)
+	if err != nil {
+		t.Fatal(err)
+	}
+	req = logical.TestRequest(t, logical.UpdateOperation, "auth/userpass/tune")
+	req.Data["user_lockout_config"] = map[string]interface{}{
+		"lockout_threshold": "3",
+	}
+	req.ClientToken = rootToken
+	resp, err = b.HandleRequest(namespace.RootContext(nil), req)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	req = logical.TestRequest(t, logical.ReadOperation, "auth/userpass/tune")
+	req.ClientToken = rootToken
+	resp, err = b.HandleRequest(namespace.RootContext(nil), req)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	if resp == nil {
+		t.Fatal("resp is nil")
+	}
+	if resp.Data["user_lockout_threshold"] != uint64(3) {
+		t.Fatalf("got: %#v expect: %#v", resp.Data["user_lockout_threshold"], uint64(3))
+	}
+
+	// tune user_lockout_config using mounts/auth/ tune with rootToken
+	// this should work as user_lockout_config needs sudo capability
+	req = logical.TestRequest(t, logical.UpdateOperation, "auth/userpass/tune")
+	req.Data["user_lockout_config"] = map[string]interface{}{
+		"lockout_threshold": "5",
+	}
+	req.ClientToken = rootToken
+	resp, err = b.HandleRequest(namespace.RootContext(nil), req)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	req = logical.TestRequest(t, logical.ReadOperation, "mounts/auth/userpass/tune")
+	resp, err = b.HandleRequest(namespace.RootContext(nil), req)
+	if err != nil {
+		t.Fatalf("err: %v", err)
+	}
+	if resp == nil {
+		t.Fatal("resp is nil")
+	}
+	if resp.Data["user_lockout_threshold"] != uint64(5) {
+		t.Fatalf("got: %#v expect: %#v", resp.Data["user_lockout_threshold"], uint64(5))
+	}
+
+	// tune user_lockout_config using mounts/auth/ tune without rootToken
+	// this should not work as user_lockout_config needs sudo capability
+	req = logical.TestRequest(t, logical.UpdateOperation, "mounts/auth/userpass/tune")
+	req.Data["user_lockout_config"] = map[string]interface{}{
+		"lockout_threshold": "5",
+	}
+	resp, err = b.HandleRequest(namespace.RootContext(nil), req)
+	if err == nil || resp == nil || !resp.IsError() || !strings.Contains(resp.Error().Error(), "tuning of user lockout configuration needs sudo capability") {
+		t.Fatalf("expected tune request to fail, but got resp: %#v, err: %s", resp, err)
+	}
+}
+
```

Comment:
Is it worth adding another test case to show what happens for a request that doesn't contain the `user_lockout_config` parameter?

Reply:
There are many tests in the code where we do not pass "user_lockout_config" to mounts tune endpoint. For better understanding, I added a test to test this separately now.